Week of July 20 (#500)
xregistry/spec: core/spec.md

- Type: URL
- Description: A server-generated unique absolute URL for an entity. This
  attribute MUST be an alternative URL for the owning entity's `self` URL,
  without any `$details` suffix. In this respect, the `shortself` URL MAY
  have `$details` or any other flags (e.g. query parameters) that are valid

>   for use on the `self` URL.

Reviewer: Okay, so `$details` can't be part of the path, but can be a query parameter, right?

Author: it can be added yes, but it's not a query parameter.
So:
```
non-shortself: http://example.com/dirs/d1/files/f1
shortself: http://example.com/r/1234
```
and
```
non-shortself: http://example.com/dirs/d1/files/f1$details
shortself: http://example.com/r/1234$details
```

make sense?

Reviewer: Okay, I *think* I get it now, but the "without any `$details` suffix" is ambiguous - I read it as being "the `shortself` mustn't have `$details` even if `self` does".

If you were to just remove "without any `$details` suffix" would we be losing information? If not, let's remove it - if so, let's clarify.

Author: see if the new text is any better

- Type: URL
- Description: A server-generated unique absolute URL for an entity. This
  attribute MUST be an alternative URL for the owning entity's `self`
  (non-`$details` suffixed) URL. When used as part of a request, the
  `shortself` URL MAY include `$details`, or any flags (e.g. query parameters),

>   that are valid for use on the `self` URL.

Reviewer: Ah, I think I see now. So the `shortself` *in the response* should never include `$details`, but a client can *add* `$details` to `shortself` and get the same effect as the longer URL.

I think a tweak is still needed, because the `shortself` URL itself *does not* include `$details`... when making a request, you're not saying "this is the `shortself` URL" you're just using a URL which has the value copied from the `shortself` attribute.

How about:

> When a client constructs a request based on the `shortself` URL, it may append `$details` or any flags (e.g. query parameters) that are valid for use on the `self` URL.

?

Author: Yup - and minor tweak to what you said though:

...but a client can add $details to shortself and get the same effect as the longer URL.
should be:
...but a client can add $details to shortself and get the same effect as the longer URL with $details added.

I assume that's what you meant.

Reviewer: Yup - basically "whatever you add to shortself will have the same effect as if you added it to self".

  attribute, with any value (even an empty string). In other words, the filter
  will only fail if the attribute has no value at all.

As specified in the [Registry Collections](#registry-collections) section,
when a query includes a filter expression, any nested xRegistry collections
in the response SHOULD result in the specification of a `<COLLECTION>url`
attribute that allows clients to easily traverse into that collection with
the appropriated sub-filter expression to yield the same results as the
original filter, but just for that nested collection. However, if the filters
results in zero entities for a particular collection due to the filter
excluding that collection entirely, rather than the `<COLLECTION>url` not
having any filter expression at all (which would result in the full collection

> being returned), servers MUST use the special filter expression `excludeall`

Reviewer: I really love the clarity of `excludeall` as a filter. Unambiguous, succinct... which genius came up with that?

;)

to indicate no results are to be returned for that URL.

For example, using HTTP:
```yaml
GET /?filter=schemagroups.schemas.name=myschema

HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8

{
  ... Registry attributes excluded for brevity ...
  "messagegroupsurl": "http://example.com/messagegroups?filter=excludeall",
  "messagegroupscount": 0,
  "schemagroupsurl": "http://example.com/schemagroups?filter=schemas.name=myschema",
  "schemagroupscount": 10
}
```

If the URL references a collection then an empty collection (`{}`) MUST be
returned. If the URL references an entity, then an error
([not_found](#not_found)) MUST be generated. If any other filter expression
appear at the same time, then an error ([bad_filter](#bad_filter)) MUST be
generated.

> 

Reviewer: Nit: double blank line

Author: fixed

> There is special `filter` value (`excludeall`) that MAY be used to indicate

Reviewer: ```suggestion
The special `filter` value `excludeall` MAY be used to indicate
```

(Just to avoid the crucial information, the value, from being parenthesized...)

Alternatively, "There is a special `filter` value" - i.e. add "a" before "special".

Author: done

that no results are to be returned from the query. This value is expected to

> only be used by server's in response to a query that produces a

Reviewer: ```suggestion
only be used by servers in response to a query that produces a
```

Author: done

> `<COLLECTION>url`

Reviewer: Incomplete sentence?

Author: forgot the trailing period. fixed
